src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -0,0 +1,176 @@
+package is.hail.io.bgen
+
+import is.hail.annotations._
+import is.hail.io.KeySerializedValueRecord
+import is.hail.io.gen.GenReport._
+import is.hail.utils._
+import is.hail.variant.{DosageGenotype, Genotype, Variant}
+
+abstract class BgenRecord extends KeySerializedValueRecord[Variant, Iterable[Genotype]] {
+  var ann: Annotation = _
+
+  def setAnnotation(ann: Annotation) {
+    this.ann = ann
+  }
+
+  def getAnnotation: Annotation = ann
+
+  override def getValue: Iterable[Genotype]
+}
+
+class BgenRecordV11(compressed: Boolean,
+  nSamples: Int,
+  tolerance: Double) extends BgenRecord {
+
+  override def getValue: Iterable[Genotype] = {
+    require(input != null, "called getValue before serialized value was set")
+
+    val bytes = if (compressed) decompress(input, nSamples * 6) else input
+
+    assert(bytes.length == nSamples * 6)
+
+    resetWarnings()
+
+    val lowerTol = (32768 * (1.0 - tolerance) + 0.5).toInt
+    val upperTol = (32768 * (1.0 + tolerance) + 0.5).toInt
+    assert(lowerTol > 0)
+
+    val noCall: Genotype = new DosageGenotype(-1, null)
+
+    new Iterable[Genotype] {
+      def iterator = new Iterator[Genotype] {
+        var i = 0
+
+        def hasNext: Boolean = i < bytes.length
+
+        def next(): Genotype = {
+          val d0 = (bytes(i) & 0xff) | ((bytes(i + 1) & 0xff) << 8)
+          val d1 = (bytes(i + 2) & 0xff) | ((bytes(i + 3) & 0xff) << 8)
+          val d2 = (bytes(i + 4) & 0xff) | ((bytes(i + 5) & 0xff) << 8)
+
+          i += 6
+
+          val dsum = d0 + d1 + d2
+          if (dsum >= lowerTol) {
+            if (dsum <= upperTol) {
+              val px =
+                if (dsum == 32768)
+                  Array(d0, d1, d2)
+                else
+                  Genotype.weightsToLinear(d0, d1, d2)
+              val gt = Genotype.unboxedGTFromLinear(px)
+              new DosageGenotype(gt, px)
+            } else {
+              setWarning(dosageGreaterThanTolerance)
+              noCall
+            }
+          } else {
+            if (dsum == 0)
+              setWarning(dosageNoCall)
+            else
+              setWarning(dosageLessThanTolerance)
+
+            noCall
+          }
+        }
+      }
+    }
+  }
+}
+
+class BgenRecordV12(compressed: Boolean, nSamples: Int, tolerance: Double) extends BgenRecord {
+  var expectedDataSize: Int = _
+  var expectedNumAlleles: Int = _
+
+  def setExpectedDataSize(size: Int) {
+    this.expectedDataSize = size
+  }
+
+  def setExpectedNumAlleles(n: Int) {
+    this.expectedNumAlleles = n
+  }
+
+  override def getValue: Iterable[Genotype] = {
+    require(input != null, "called getValue before serialized value was set")
+
+    val bytes = if (compressed) decompress(input, expectedDataSize) else input
+    val nRow = (bytes(0) & 0xff) | ((bytes(1) & 0xff) << 8) | ((bytes(2) & 0xff) << 16) | ((bytes(3) & 0xff) << 24)
```

Comment:
You might write some utility functions for grabbing out little endian shorts, ints from the byte array.